utils/json.go

```diff
@@ -39,45 +40,81 @@ type JSONFragment []byte
 
 // Resolve resolves the given key when this JSON fragment is referenced in an expression
 func (j JSONFragment) Resolve(key string) interface{} {
-	_, err := strconv.Atoi(key)
+	_, isIndex := strconv.Atoi(key)
 
 	// this is a numerical index, convert to jsonparser format
-	if err == nil {
+	if isIndex == nil {
 		jIdx := "[" + key + "]"
 		val, valType, _, err := jsonparser.Get(j, jIdx)
 		if err == nil {
-			if err == nil {
-				if valType == jsonparser.String {
-					strVal, err := jsonparser.ParseString(val)
-					if err == nil {
-						return strVal
-					}
-				}
-				return JSONFragment(val)
-			}
+			return jsonTypeToXAtom(val, valType)
 		}
 	}
+
 	val, valType, _, err := jsonparser.Get(j, key)
 	if err != nil {
 		return fmt.Errorf("no such variable: %s", key)
 	}
 
-	if valType == jsonparser.String {
-		strVal, err := jsonparser.ParseString(val)
-		if err == nil {
-			return strVal
-		}
-	}
-	return JSONFragment(val)
+	return jsonTypeToXAtom(val, valType)
 }
 
-// String returns the string representation of this JSON, which is just the JSON itself
+// Atomize is called when this object needs to be reduced to a primitive
 func (j JSONFragment) Atomize() interface{} {
 	return string(j)
 }
 
-var _ VariableAtomizer = EmptyJSONFragment
-var _ VariableResolver = EmptyJSONFragment
+var _ Atomizable = EmptyJSONFragment
+var _ Resolvable = EmptyJSONFragment
+
+// JSONArray is a JSON fragment containing an array
+type JSONArray JSONFragment
+
+// Index is called when this object is indexed into in an expression
+func (j JSONArray) Index(index int) interface{} {
+	val, valType, _, err := jsonparser.Get(j, fmt.Sprintf("[%d]", index))
+	if err != nil {
+		return err
+	}
+	return jsonTypeToXAtom(val, valType)
+}
+
+// Length is called when the length of this object is requested in an expression
+func (j JSONArray) Length() int {
+	length := 0
+	jsonparser.ArrayEach(j, func(value []byte, dataType jsonparser.ValueType, offset int, err error) {
+		length++
+	})
+	return length
+}
+
+var _ Indexable = JSONArray{}
+
+func jsonTypeToXAtom(data []byte, valType jsonparser.ValueType) interface{} {
```

Comment:
@nicpottier now any time we resolve something out of a `JSONFragment`, we see if we can convert it to one of our atomic types. `JSONArray` is a special case of a fragment.. which is an array and thus can be indexed into and has length.